feat(delegation) Mark self delegation

TimmyExogenous opened the pull request:
## Description

This PR implements the functionality regarding marking the operator as the staker's owner for the EVM-type client chains. 

* provide `markSelfDelegatedOperator` and `unmarkSelfDelegatedOperator` in the delegation precompile
* add a command line `QueryMarkedOperatorByStaker` to query the related information
* remove the useless field `operator_amount` in the `OperatorAssetInfo`, which will cause an incompatible state. So you need to reset the node when using the `local_node.sh`
* fix the issue regarding the error handling in the precompile.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added new functions to associate and dissociate operators with stakers.
  - Enhanced contract capabilities to manage operator-staker relationships.
  - Implemented a method for updating parameters with strict validation.
  - Introduced new error constants for improved error handling.

- **Bug Fixes**
  - Updated error handling for better clarity and consistency.

- **Refactor**
  - Adjusted parameter names for consistency across delegation operations.
  - Streamlined command structure for dogfood-related CLI commands.
  - Modified data structures to support new associations in genesis state.

- **Tests**
  - Updated and added new test cases to ensure correctness of delegation and operator functionalities.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
feat(dogfood): implement update params msg

MaxMustermann2 committed:
fix(dogfood): remove deprecated x/params

MaxMustermann2 committed:
fix(app): do not hardcode authority

MaxMustermann2 committed:
fix(dogfood): set `post` to `tx` not get

MaxMustermann2 committed:
fix(dogfood): remove typo

MaxMustermann2 committed:
refactor(dogfood): improve msg_server readability

TimmyExogenous committed:
fix the issues identified by code review in#PR116

TimmyExogenous committed:
fix the issues identified by code review in#PR116

TimmyExogenous committed:
Add self-delegation marking for EVM client chains

TimmyExogenous committed:
Add self-delegation marking for EVM client chains

TimmyExogenous committed:
add code comments for the protobuf

TimmyExogenous committed:
add code comments for the protobuf

MaxMustermann2 commented on precompiles/delegation/IDelegation.sol:
```diff
@@ -52,5 +52,29 @@ interface IDelegation {
         uint256 opAmount
     ) external returns (bool success);
+
+    /// TRANSACTIONS
```
This line is superfluous; please remove.

TimmyExogenous replied:
Thanks, I will remove them.

MaxMustermann2 commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +96,81 @@ func (p Precompile) UndelegateFromThroughClientChain(
+	if err != nil {
+		return nil, errorsmod.Wrap(err, exocmn.ErrContractCaller)
+	}
+	if len(args) != 3 {
```
```suggestion
	if len(args) != len(p.ABI.Methods[MethodMarkSelfDelegatedOperator].Inputs) {
```

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +96,81 @@ func (p Precompile) UndelegateFromThroughClientChain(
+	if err != nil {
+		return nil, errorsmod.Wrap(err, exocmn.ErrContractCaller)
+	}
+	if len(args) != 2 {
```
```suggestion
	if len(args) != len(p.ABI.Methods[MethodUnmarkSelfDelegatedOperator].Inputs) {
```

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/delegation/keeper/delegation.go:
```diff
@@ -63,10 +65,20 @@ func (k *Keeper) delegateTo(
+	if err != nil {
+		return err
+	}
+	if operator != "" && operator == params.OperatorAddress.String() {
```
Since `params.OperatorAddress.String()` is not `""`, the first check is unnecessary.
```suggestion
	if operator == params.OperatorAddress.String() {
```

TimmyExogenous replied:
Thanks

MaxMustermann2 commented on x/delegation/keeper/delegation.go:
```diff
@@ -136,3 +148,107 @@ func (k *Keeper) UndelegateFrom(ctx sdk.Context, params *delegationtype.Delegati
+		return err
+	}
+
+	if oldOperator != "" && oldOperator == operatorAddress.String() {
```
I don't think we should support reassigning a staker's self delegation to another operator.

TimmyExogenous replied:
Yeah, it's fine not to support reassigning a self-delegation. The staker can correct the marking mistake by unmarking and marking the operator again. I will remove the reassigning logic and check if there has been a marked operator before marking.

MaxMustermann2 commented on x/delegation/keeper/delegation.go:
```diff
@@ -136,3 +148,107 @@ func (k *Keeper) UndelegateFrom(ctx sdk.Context, params *delegationtype.Delegati
+	if err != nil {
+		return err
+	}
+
```
We should verify via signing the ownership of the `stakerID` by the `operatorAddress`

TimmyExogenous replied:
The assets are controlled by the staker, so the marking should check the signature of the staker. This function will use the input staker address from the Exocore gateway directly. The signature verification will be performed when processing the EVM transaction. The gateway contract should use the recovered caller address as the input staker address when calling the precompiled contract. This documentation provides more information.
https://www.notion.so/exocore/Marking-self-delegation-and-the-staking-compatibility-for-the-other-non-EVM-client-chains-48cfa40ec5eb41b0a2a484c1bda0f65b?pvs=4

MaxMustermann2 committed:
feat(dogfood): introduce min self delegation param

MaxMustermann2 committed:
feat: implement min self delegation for dogfood

MaxMustermann2 committed:
local: set 100 min self delegation

MaxMustermann2 committed:
feat(dogfood): allow edit min self delegation

MaxMustermann2 commented:
Test object: b4d6d615b2903972eb65753673e2424904640100 + c92c10f5971d4c14b03ccb1a8771bdd5067d2df0 (from #134) 

1. Start the network using `local_node.sh` which has a validator with 5000 vote power
1. Wait for the next minute and see a consensus failure in the logs to indicate the "applying the validator changes would result in empty set"

This happens because the delegation of 5000 is not specified to be the self delegation in the genesis file; it is implicitly assumed to be that from a different staker. To fix this, I will rework some of the contents of `x/delegation` genesis to make marking self-stake possible.

MaxMustermann2 committed:
chore: lint and pass test

TimmyExogenous committed:
correct GetAllClientChainInfo in the IAssets interface

TimmyExogenous commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -124,24 +124,16 @@
-  string operator_unbonding_amount = 4
+  string operator_unbonding_amount = 3
   [
     (cosmos_proto.scalar) = "cosmos.Int",
```
I didn't rename `operator_unbonding_amount` to `wait_unbonding_amount`, because the above useless field has been removed. Additionally, the field `operator_unbonding_amount` is useless, so I will remove it in the subsequent commit.

TimmyExogenous commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -160,8 +152,8 @@
-  //todo: the field is used to mark operator's own asset share and is not temporarily used now
-  string operator_share = 6
+  // the field is used to mark operator's own asset share.
+  string operator_share = 5
   [
     (cosmos_proto.scalar) = "cosmos.Dec",
```
Because some useless fields are removed, this isn't an issue regarding field renaming.

TimmyExogenous commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +96,81 @@ func (p Precompile) UndelegateFromThroughClientChain(
+	_ common.Address,
+	contract *vm.Contract,
+	_ vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	// check the invalidation of caller contract
+	err := p.assetsKeeper.CheckExocoreGatewayAddr(ctx, contract.CallerAddress)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, exocmn.ErrContractCaller)
+	}
+	if len(args) != 3 {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, 3, len(args))
+	}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	staker, ok := args[1].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 1, "string", args[1])
+	}
+	if !common.IsHexAddress(staker) {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidEVMAddr, staker)
+	}
+
+	operator, ok := args[2].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, "string", args[2])
+	}
+	operatorAccAddr, err := sdk.AccAddressFromBech32(operator)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, fmt.Sprintf("error occurred when parse the operator address from Bech32,the operator is:%s", operator))
+	}
+	err = p.delegationKeeper.MarkSelfDelegatedOperator(ctx, uint64(clientChainID), operatorAccAddr, common.HexToAddress(staker))
+	if err != nil {
+		return nil, err
+	}
+	return method.Outputs.Pack(true)
+}
```
Thanks

TimmyExogenous replied:
Thanks, I will use `xerrors.Errorf` to handle all errors, because there isn't a module name in the precompiled contract.

TimmyExogenous commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +96,81 @@
+	return method.Outputs.Pack(true)
+}
+
+func (p Precompile) UnMarkSelfDelegatedOperator(
+	ctx sdk.Context,
+	_ common.Address,
+	contract *vm.Contract,
+	_ vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	// check the invalidation of caller contract
+	err := p.assetsKeeper.CheckExocoreGatewayAddr(ctx, contract.CallerAddress)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, exocmn.ErrContractCaller)
+	}
+	if len(args) != 2 {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, 2, len(args))
+	}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	staker, ok := args[1].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 1, "string", args[1])
+	}
+	if !common.IsHexAddress(staker) {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidEVMAddr, staker)
+	}
+
+	err = p.delegationKeeper.UnmarkSelfDelegatedOperator(ctx, uint64(clientChainID), common.HexToAddress(staker))
+	if err != nil {
+		return nil, err
+	}
+	return method.Outputs.Pack(true)
+}
```
Thanks, I will use `xerrors.Errorf` to handle all errors, because there isn't a module name in the precompiled contract.

TimmyExogenous commented on precompiles/common/error.go:
```diff
@@ -13,4 +13,6 @@ const (
 
 	ErrInvalidNameLength = "nil name or too long for chain or token,value:%s,actualLength:%d,max:%d"
+
+	ErrInvalidEVMAddr = "the address isn't an invalid EVM address, addr:%s"
```
Thanks

TimmyExogenous committed:
fix the issues identified in the code comments：
1. rename the `markSelfDelegation` to `associateOperatorWithStaker`
2. use `bytes` to replace `string` for the input parameter of `associateOperatorWithStaker`
3. use `xerrors` to handle all errors in the precompiled contracts
4. remove operator reassignment regarding self-delegation association
5. remove the useless field `operatorUnbondingAmount` in the `OperatorAssetInfo`

TimmyExogenous committed:
fix the lint error regarding too long rpc in the protobuf

MaxMustermann2 committed:
feat(delegation): add association data to genesis

MaxMustermann2 committed:
fix(dogfood): log Info not Error

debugging remnant

MaxMustermann2 committed:
feat(delegation): implement association at genesis

MaxMustermann2 commented:
Test object: 0e4ef8fdfc919d8f5b5ebcb3bcb09e32f0f389ad

1. Test that consensus is not lost after one epoch. Start `local_node.sh` and wait for one epoch (minute) to go ahead. Test passes.

2. Test that consensus is lost, if the `min_self_delegation` is increased to a value greater than the single validator's vote power.

Apply this patch to change the `min_self_delegation` on the fly.
```patch
diff --git a/x/dogfood/keeper/msg_server.go b/x/dogfood/keeper/msg_server.go
index 69c336ed..97566a15 100644
--- a/x/dogfood/keeper/msg_server.go
+++ b/x/dogfood/keeper/msg_server.go
@@ -4,11 +4,8 @@ import (
 	"context"
 	"strings"
 
-	errorsmod "cosmossdk.io/errors"
-
 	"github.com/ExocoreNetwork/exocore/x/dogfood/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
 	epochstypes "github.com/evmos/evmos/v14/x/epochs/types"
 )
 
@@ -29,13 +26,13 @@ func (k Keeper) UpdateParams(
 	ctx context.Context, msg *types.MsgUpdateParams,
 ) (*types.MsgUpdateParamsResponse, error) {
 	c := sdk.UnwrapSDKContext(ctx)
-	if k.authority != msg.Authority {
-		return nil, errorsmod.Wrapf(
-			govtypes.ErrInvalidSigner,
-			"invalid authority; expected %s, got %s",
-			k.authority, msg.Authority,
-		)
-	}
+	// if k.authority != msg.Authority {
+	// 	return nil, errorsmod.Wrapf(
+	// 		govtypes.ErrInvalidSigner,
+	// 		"invalid authority; expected %s, got %s",
+	// 		k.authority, msg.Authority,
+	// 	)
+	// }
 	prevParams := k.GetDogfoodParams(c)
 	nextParams := msg.Params
 	logger := k.Logger(c)
```

Run `local_node.sh` from scratch, and wait for an epoch to pass. Send a transaction to alter the minimum self delegation: `exocored tx dogfood update-params --min-self-delegation 5001 --from local_funded_account --home ~/.tmp-exocored --gas-prices 875000000aexo`. Wait for the transaction to execute and the current epoch to end. Test passes.

```
6:46PM ERR CONSENSUS FAILURE!!! err="failed to apply block; error commit failed for application: error changing vali
dator set: applying the validator changes would result in empty set" module=consensus server=node stack="goroutine 4
5 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x5e\ngithub.com/cometbft/cometbft/consensus.(*Sta
te).receiveRoutine.func2()\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:736 +0x46\npanic({0x2c791a0?, 
0xc0057af520?})\n\truntime/panic.go:914 +0x21f\ngithub.com/cometbft/cometbft/consensus.(*State).finalizeCommit(0xc00
167a000, 0x15)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1719 +0xec5\ngithub.com/cometbft/cometbft/
consensus.(*State).tryFinalizeCommit(0xc00167a000, 0x15)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:
1619 +0x2f6\ngithub.com/cometbft/cometbft/consensus.(*State).enterCommit.func1()\n\tgithub.com/cometbft/cometbft@v0.
37.4/consensus/state.go:1554 +0x9c\ngithub.com/cometbft/cometbft/consensus.(*State).enterCommit(0xc00167a000, 0x15, 
0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1592 +0xc5b\ngithub.com/cometbft/cometbft/consensus.
(*State).addVote(0xc00167a000, 0xc00566e0a0, {0x0, 0x0})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:
2223 +0x18ed\ngithub.com/cometbft/cometbft/consensus.(*State).tryAddVote(0xc00167a000, 0xc00566e0a0, {0x0?, 0x4df290
?})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2012 +0x26\ngithub.com/cometbft/cometbft/consensus.(*
State).handleMsg(0xc00167a000, {{0x416d080, 0xc0050a4040}, {0x0, 0x0}})\n\tgithub.com/cometbft/cometbft@v0.37.4/cons
ensus/state.go:865 +0x3ce\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine(0xc00167a000, 0x0)\n\tgith
ub.com/cometbft/cometbft@v0.37.4/consensus/state.go:792 +0x4bc\ncreated by github.com/cometbft/cometbft/consensus.(*
State).OnStart in goroutine 1\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:383 +0x10c\n"
```

MaxMustermann2 commented on x/delegation/types/genesis.go:
```diff
@@ -106,5 +108,35 @@ func (gs GenesisState) Validate() error {
 			}
 		}
 	}
+	// for associations, one stakerID can be associated only with one operator.
+	// but one operator may have multiple stakerIDs associated with it.
+	associatedStakerIDs := make(map[string]struct{}, len(gs.Associations))
+	for _, association := range gs.Associations {
+		// check operator address
+		if _, err := sdk.AccAddressFromBech32(association.Operator); err != nil {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData,
+				"invalid operator address for operator %s", association.Operator,
+			)
+		}
+		// check staker address
+		if _, _, err := assetstypes.ValidateID(
+			association.StakerID, true, true,
+		); err != nil {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData, "invalid staker ID %s: %s", association.StakerID, err,
+			)
+		}
+		// check for duplicate stakerIDs
+		if _, ok := associatedStakerIDs[association.StakerID]; ok {
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData, "duplicate staker ID %s", association.StakerID,
+			)
+		}
+		associatedStakerIDs[association.StakerID] = struct{}{}
+		// we don't check that this `association.stakerID` features in `gs.Delegations`,
+		// because we allow the possibility of a staker without any delegations to be associated
+		// with an operator.
+	}
```
Sure, create the issue and assign me.

MaxMustermann2 commented on x/delegation/keeper/delegation.go:
```diff
@@ -136,3 +148,109 @@ func (k *Keeper) UndelegateFrom(ctx sdk.Context, params *delegationtype.Delegati
+	}
+
+	stakerID, _ := assetstype.GetStakeIDAndAssetID(clientChainID, stakerAddress[:], nil)
+	associatedOperator, err := k.GetSelfDelegatedOperator(ctx, stakerID)
+	if err != nil {
+		return err
+	}
+	if associatedOperator != "" {
+		return delegationtype.ErrOperatorAlreadyAssociated
+	}
```
This isn't code golf. We prefer code to be readable.

MaxMustermann2 commented on x/delegation/keeper/delegation.go:
```diff
@@ -63,10 +65,20 @@
 		TotalShare:  share,
-	})
+	}
+	// Check if the staker belongs to the delegated operator. Increase the operator's share if yes.
+	operator, err := k.GetSelfDelegatedOperator(ctx, stakerID)
+	if err != nil {
+		return err
+	}
+	if operator == params.OperatorAddress.String() {
+		deltaOperatorAsset.OperatorShare = share
+	}
```
This isn't code golf. We prefer code to be readable.

MaxMustermann2 commented on x/delegation/keeper/delegation.go:
```diff
@@ -136,3 +148,109 @@
+	stakerAddress common.Address,
+) error {
+	stakerID, _ := assetstype.GetStakeIDAndAssetID(clientChainID, stakerAddress[:], nil)
+	associatedOperator, err := k.GetSelfDelegatedOperator(ctx, stakerID)
+	if err != nil {
+		return err
+	}
+	if associatedOperator == "" {
+		return delegationtype.ErrNoAssociatedOperatorByStaker
+	}
```
This isn't code golf. We prefer code to be readable.

MaxMustermann2 commented on x/dogfood/keeper/msg_server.go:
```diff
@@ -15,3 +23,88 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+		)
+		nextParams.AssetIDs = prevParams.AssetIDs
+	}
+	if nextParams.MinSelfDelegation.IsNil() || nextParams.MinSelfDelegation.IsNegative() {
+		logger.Info(
+			"UpdateParams",
+			"overriding MinSelfDelegation with value", prevParams.MinSelfDelegation,
+		)
+		nextParams.MinSelfDelegation = prevParams.MinSelfDelegation
+	}
```
Did you consider that the minimum self delegation is allowed to be 0 before you made that comment?

adu-web3 commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +93,88 @@ func (p Precompile) UndelegateFromThroughClientChain(
+	if err != nil {
+		return nil, xerrors.Errorf("error occurred when parse the operator address from Bech32,the operator is:%s, error:%s ", operator, err.Error())
+	}
+	err = p.delegationKeeper.AssociateOperatorWithStaker(ctx, uint64(clientChainID), operatorAccAddr, common.Address(staker))
```
should we check some circumstances before associating operator with staker?

1. the client chain id should have been registered
2. the stakerID should not have been associated before

btw, for current design, one operator could be associated to several stakes for self-delegate, is this intended?

TimmyExogenous replied:
Thanks. The client chain ID should be checked. I will add the check in the native module. It has already been checked whether the stakerID was associated before its association in the native module.
Additionally, it's intended that one operator can be associated with several stakers.

TimmyExogenous committed:
check if the client chain has been registered before associating operator

bwhour commented on precompiles/assets/assets.go:
```diff
@@ -3,10 +3,9 @@ package assets
 	"math/big"
 
-	sdk "github.com/cosmos/cosmos-sdk/types"
+	"golang.org/x/xerrors"
```
can you explain why use this third package not use standard library fmt ? do you want stack trace information ?

TimmyExogenous replied:
It's a Golang extended library maintained by the Go team. It seems to provide richer functionality, but I didn't consider whether it's better to use `fmt` here. I can change it to `fmt` if you think it's better.

bwhour commented on x/delegation/client/cli/query.go:
```diff
@@ -210,3 +211,36 @@ func QueryUndelegationHoldCount() *cobra.Command {
+// QueryAssociatedOperatorByStaker queries the operator owner of the specified staker
+func QueryAssociatedOperatorByStaker() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "QueryAssociatedOperatorByStaker stakerID",
```
QueryAssociatedOperatorByStaker [stakerID]

TimmyExogenous replied:
Thanks

bwhour commented on x/delegation/keeper/genesis.go:
```diff
@@ -50,6 +50,22 @@ func (k Keeper) InitGenesis(
+		stakerID := association.StakerID
+		operatorAddress := association.Operator
+		// #nosec G703 // already validated
+		stakerAddress, lzID, _ := assetstype.ParseID(stakerID)
```
lzID is not clear for reader, let's add todo to replace this with clientChainID.

TimmyExogenous replied:
It will be renamed to `clientChainID`

TimmyExogenous committed:
fix the issues idenfied by code review:
1. use `fmt.Errorf` to replace `xerrors.Errorf` in the precompiles
2. rename the `lzID` to `clientChainID`
3. use `<>` to enclose the command line parameters

TimmyExogenous commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -124,24 +124,8 @@
-
-  // operator_unbonding_amount is the amount that is owned by operator itself and waiting for unbonding.
-  string operator_unbonding_amount = 4
+  string wait_unbonding_amount = 2
```
It's necessary to mark the custom name, because the default name is fine when generating the go file.

TimmyExogenous commented on proto/exocore/assets/v1/tx.proto:
```diff
@@ -160,8 +144,8 @@
-  //todo: the field is used to mark operator's own asset share and is not temporarily used now
-  string operator_share = 6
+  // the field is used to mark operator's own asset share.
+  string operator_share = 4
```
It's necessary to mark the custom name, because the default name is fine when generating the go file.

TimmyExogenous commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +92,88 @@ func (p Precompile) UndelegateFromThroughClientChain(
+	args []interface{},
+) ([]byte, error) {
+	// check the invalidation of caller contract
+	err := p.assetsKeeper.CheckExocoreGatewayAddr(ctx, contract.CallerAddress)
+	if err != nil {
+		return nil, fmt.Errorf(exocmn.ErrContractCaller, err.Error())
+	}
+
+	inputsLen := len(p.ABI.Methods[MethodAssociateOperatorWithStaker].Inputs)
+	if len(args) != inputsLen {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, inputsLen, len(args))
+	}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	staker, ok := args[1].([]byte)
+	if !ok || staker == nil {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, "[]byte", args[1])
+	}
+	// TODO: In the future, the check should be the same as it is in delegation if using LayerZero to route the
+	// message for non-EVM client chains, such as Solana.
+	if len(staker) != common.AddressLength {
+		return nil, fmt.Errorf(exocmn.ErrInvalidEVMAddr, staker)
+	}
+
+	operator, ok := args[2].([]byte)
+	if !ok || operator == nil {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 2, "[]byte", args[2])
+	}
+	operatorAccAddr, err := sdk.AccAddressFromBech32(string(operator))
+	if err != nil {
+		return nil, fmt.Errorf("error occurred when parse the operator address from Bech32,the operator is:%s, error:%s ", operator, err.Error())
+	}
+	err = p.delegationKeeper.AssociateOperatorWithStaker(ctx, uint64(clientChainID), operatorAccAddr, common.Address(staker))
+	if err != nil {
+		return nil, err
+	}
+	return method.Outputs.Pack(true)
+}
```
thanks

TimmyExogenous replied:
It will be changed in the Unjail PR.

TimmyExogenous commented on x/delegation/keeper/delegation.go:
```diff
@@ -136,3 +148,112 @@ func (k *Keeper) UndelegateFrom(ctx sdk.Context, params *delegationtype.Delegati
+		return delegationtype.ErrOperatorNotExist
+	}
+
+	stakerID, _ := assetstype.GetStakeIDAndAssetID(clientChainID, stakerAddress[:], nil)
+	associatedOperator, err := k.GetSelfDelegatedOperator(ctx, stakerID)
+	if err != nil {
+		return err
+	}
+	if associatedOperator != "" {
+		return delegationtype.ErrOperatorAlreadyAssociated
+	}
```
The returned errors are different.

TimmyExogenous commented on precompiles/delegation/tx.go:
```diff
@@ -85,3 +92,88 @@
+	return method.Outputs.Pack(true)
+}
+
+func (p Precompile) DissociateOperatorFromStaker(
+	ctx sdk.Context,
+	_ common.Address,
+	contract *vm.Contract,
+	_ vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	// check the invalidation of caller contract
+	err := p.assetsKeeper.CheckExocoreGatewayAddr(ctx, contract.CallerAddress)
+	if err != nil {
+		return nil, fmt.Errorf(exocmn.ErrContractCaller, err.Error())
+	}
+	inputsLen := len(p.ABI.Methods[MethodDissociateOperatorFromStaker].Inputs)
+	if len(args) != inputsLen {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, inputsLen, len(args))
+	}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	staker, ok := args[1].([]byte)
+	if !ok || staker == nil {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, "[]byte", args[1])
+	}
+	// TODO: In the future, the check should be the same as it is in delegation if using LayerZero to route the
+	// message for non-EVM client chains, such as Solana.
+	if len(staker) != common.AddressLength {
+		return nil, fmt.Errorf(exocmn.ErrInvalidEVMAddr, staker)
+	}
+
+	err = p.delegationKeeper.DissociateOperatorFromStaker(ctx, uint64(clientChainID), common.Address(staker))
+	if err != nil {
+		return nil, err
+	}
+	return method.Outputs.Pack(true)
+}
```
There isn't this error in the `DissociateOperatorFromStaker`

cloud8little commented:
Test passed for following scenarioes.

1. set minimum self delegation in genesis file, when new validator has self delegation less than minumum self delegation, the new validator will not able to join consensus.
2. set minimum self delegation in genesis file, when new validator has self delegation equal or more than minumum self delegation, the new validator will be able to join consensus.
3. After the network start to work, when update self delegation params to be greater than one of the self delegation of validators, if the validator voting power is less than 2/3 total voting power, the validator is kicked out of validator board.
4. After the network start to work, when update self delegation params to be greater than one of the self delegation of validators, if the validator voting power is greater than 2/3 total voting power, the whole network stops.
